### PR TITLE
feat: create initramfs crate

### PR DESCRIPTION
**Pull request by Roxxas96**
Implement a crate `initramfs` that pull docker images and wrap them into an initramfs that can be loaded into `lumper`.

### CODE REVIEW

File: initramfs/src/image.rs [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+        // TODO Use config for auth URL
+        let res = reqwest::get(format!(
+            "https://auth.docker.io/token?service=registry.docker.io&scope=repository:{}:pull",
+            self.repository
+        ))
+        .await?;
+
+        if res.status() != 200 {
+            error!("Failed to get auth token : \n{:?}", res);
+            return Err(anyhow!("Failed to get auth token"));
```
Comment (sameo) [inline initramfs/src/image.rs]: You could add the HTTP result as part of the anyhow error as well.

File: initramfs/src/image.rs [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+                "https://registry-1.docker.io/v2/{}/manifests/{}",
+                self.repository, self.tag
+            ))
+            .bearer_auth(self.token.clone())
+            .send()
+            .await?;
+
+        if res.status() != 200 {
+            error!("Failed to get manifest : \n{:?}", res);
+            return Err(anyhow!("Failed to get manifest"));
```
Comment (sameo) [inline initramfs/src/image.rs]: ditto

File: initramfs/src/image.rs [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+        let res = reqwest::Client::new()
+            .get(format!(
+                "https://registry-1.docker.io/v2/{}/manifests/{}",
+                self.repository, self.tag
+            ))
+            .bearer_auth(self.token.clone())
+            .send()
+            .await?;
+
+        if res.status() != 200 {
```
Comment (sameo) [inline initramfs/src/image.rs]: you only need to check for `res.is_success()`.

File: initramfs/src/image.rs [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+    }
+
+    pub async fn pull_layers(&mut self) -> Result<Vec<Archive<GzDecoder<Cursor<Bytes>>>>> {
+        self.get_layers().await?;
+
+        // For every layer, pull it and add it to the archive list
+        let mut archives = Vec::new();
+        for layer in self.layers.iter() {
+            info!("Pulling layer {:?}", layer.blobSum);
+            let archive = self.pull_layer(layer).await.unwrap();
```
Comment (sameo) [inline initramfs/src/image.rs]: Please avoid `unwrap`ing and propagate a proper error instead.

File: initramfs/src/main.rs [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+#[derive(Parser, Debug)]
+#[command(author, version, about, long_about = None)]
+struct Args {
+    #[arg(short, long)]
+    image: String,
+}
+
+#[tokio::main]
+async fn main() {
```
Comment (sameo) [inline initramfs/src/main.rs]: Your main should return a proper Result, and avoid the following `unwrap`s.

File: initramfs/src/filesystem.rs [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+                let mut entry = entry.unwrap();
+                let path = entry.header().path().unwrap().to_str().unwrap().to_string();
+
+                // If the file already exists, we don't need to add it again
+                if self.files.contains_key(&path) {
+                    continue;
+                }
+
+                // Create a new cpio builder for the file
+                let builder = NewcBuilder::new(entry.header().path().unwrap().to_str().unwrap())
```
Comment (RedbeanGit) [inline initramfs/src/filesystem.rs]: You could reuse the `path` variable defined above

File: initramfs/src/filesystem.rs [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+                    .uid(entry.header().uid().unwrap() as u32)
```
Comment (RedbeanGit) [inline initramfs/src/filesystem.rs]: As you get the header several times, maybe you should create a variable for that

File: initramfs/src/filesystem.rs [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+        }
+    }
+
+    pub fn parse_files(&mut self, archives: Vec<Archive<GzDecoder<Cursor<Bytes>>>>) {
+        // We need to reverse the layers so that the last layer is the first one in the cpio
+        for mut archive in archives.into_iter().rev() {
+            // Iterate over all the files in the layer
+            for entry in archive.entries().unwrap() {
+                let mut entry = entry.unwrap();
+                let path = entry.header().path().unwrap().to_str().unwrap().to_string();
```
Comment (RedbeanGit) [inline initramfs/src/filesystem.rs]: I think you should prevent using `unwrap` and just return an error

File: initramfs/src/filesystem.rs [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+                self.files.insert(path, (builder, Cursor::new(contents)));
+            }
+        }
+
+        debug!("Successfully parsed all layers");
+    }
+
+    pub fn write_cpio(&mut self, file_name: &str) {
+        // Write the cpio to disk
+        let archive = File::create(file_name).unwrap();
```
Comment (RedbeanGit) [inline initramfs/src/filesystem.rs]: idem, you should prevent using unwrap and handle errors gracefully

**Review by RedbeanGit — CHANGES_REQUESTED** [review]

File: initramfs/src/filesystem.rs (hunk 9 above)
Comment (Roxxas96) [reply initramfs/src/filesystem.rs]: I fixed this while force pushing for the last time

File: initramfs/src/filesystem.rs (hunk 8 above)
Comment (Roxxas96) [reply initramfs/src/filesystem.rs]: I fixed this while force pushing for the last time

File: initramfs/src/filesystem.rs (hunk 7 above)
Comment (Roxxas96) [reply initramfs/src/filesystem.rs]: I fixed this while force pushing for the last time

File: initramfs/src/filesystem.rs (hunk 6 above)
Comment (Roxxas96) [reply initramfs/src/filesystem.rs]: I fixed this while force pushing for the last time

File: initramfs/src/main.rs (hunk 5 above)
Comment (Roxxas96) [reply initramfs/src/main.rs]: I fixed this while force pushing for the last time

File: initramfs/src/image.rs (hunk 4 above)
Comment (Roxxas96) [reply initramfs/src/image.rs]: I fixed this while force pushing for the last time

File: initramfs/src/image.rs (hunk 3 above)
Comment (Roxxas96) [reply initramfs/src/image.rs]: I fixed this while force pushing for the last time

File: initramfs/src/image.rs (hunk 2 above)
Comment (Roxxas96) [reply initramfs/src/image.rs]: I fixed this while force pushing for the last time

File: initramfs/src/image.rs (hunk 1 above)
Comment (Roxxas96) [reply initramfs/src/image.rs]: I fixed this while force pushing for the last time

File: initramfs/src/image.rs [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+            token: "".to_string(),
+            layers: Vec::new(),
+        };
+
+        debug!("Successfully parsed image : {:?}", image);
+
+        Ok(image)
+    }
+
+    async fn get_auth_token(&mut self) -> Result<()> {
```
Comment (sameo) [inline initramfs/src/image.rs]: nitty nit: I'd call that one `fetch_auth_token`to avoid confusing it with a getter.

**Review by sameo — COMMENTED** [review]
Just a tiny nit, this looks pretty good to me.

File: initramfs/src/image.rs (hunk 10 above)
Comment (Roxxas96) [reply initramfs/src/image.rs]: Ok. I changed it in the last force push !

**Review by sameo — COMMENTED** [review]
Gentlemen, this PR looks really good, but your git history is a mess. I don't care that at some point you fixed a typo or that you decided to remove some init file. I want to see the structure of your feature, and so most people reading your project git history will. Incidentally, logically articulating your PRs allow for much easier reviews. If you take your time to split your 500 lines of code into 3-4 logically bound commits, this make me (as a reviewer) feel confident that you actually put some thoughts into this work.

So please rework this PR into e.g. 4 commits. This should be easy because your code is actually well architectured (Which is not obvious when I only read the git history):

1. Your tiny http client request.
2. The `Registry` module
3. The `Image` module
4. `main.rs`

File: initramfs/src/image.rs [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+            tag: tag.to_string(),
+            layers: Vec::new(),
+        };
+
+        debug!("Successfully parsed image : {:?}", image.get_name());
+
+        Ok(image)
+    }
+
+    pub fn get_name(&self) -> &str {
```
Comment (sameo) [inline initramfs/src/image.rs]: getter in Rust are usually not prefixed with `get`. Please rename to `name`

File: initramfs/src/image.rs [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+        &self.name
+    }
+
+    pub fn get_tag(&self) -> &str {
```
Comment (sameo) [inline initramfs/src/image.rs]: Ditto. s/get_tag/tag/

File: initramfs/src/image.rs (hunk 11 above)
Comment (RedbeanGit) [reply initramfs/src/image.rs]: fixed!

File: initramfs/src/image.rs (hunk 12 above)
Comment (RedbeanGit) [reply initramfs/src/image.rs]: fixed too!

**Review by sameo — APPROVED** [review]

**Pull request merged**